src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala

```diff
@@ -532,41 +584,103 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
         allVertices.map { v => v(2) }.sum / allVertices.size
       )
     } else {
-      Array(
-        // Factor of 3 comes from: centroid / (2.0 * (volume/6.0))
-        3.0 * centroidX / totalVolume,
-        3.0 * centroidY / totalVolume,
-        3.0 * centroidZ / totalVolume
-      )
+      val increments: Seq[(Double, Double, Double)] = faces.flatMap { face =>
+        vertices(face).map { faceVertex =>
+          val vertexIndex = vertices(face).indexOf(faceVertex)
+
+          // Tetrahedra are constructed using three vertices at a time, so stop
+          // mapping beyond 3 to last vertex
+          if ( vertexIndex < vertices(face).size - 2 ) {
+            val Jacobian = DenseMatrix.zeros[Double](3, 3)
+            // Translate vertex coordinates from global to local where center of mass is local origin
+            val vertex1 = DenseVector[Double](vertices(face)(0)(0) - centerX,
+              vertices(face)(0)(1) - centerY,
+              vertices(face)(0)(2) - centerZ)
+            val vertex2 = DenseVector[Double](vertices(face)(vertexIndex + 1)(0) - centerX,
+              vertices(face)(vertexIndex + 1)(1) - centerY,
+              vertices(face)(vertexIndex + 1)(2) - centerZ)
+            val vertex3 = DenseVector[Double](vertices(face)(vertexIndex + 2)(0) - centerX,
+              vertices(face)(vertexIndex + 2)(1) - centerY,
+              vertices(face)(vertexIndex + 2)(2) - centerZ)
+
+            // Initialize Jacobian
```

Comment:
I think you can initialize the Jacobian a little more cleanly. Probably you can write something like
```scala
val Jacobian = DenseMatrix[Double](vertex1, vertex2, vertex3)
```